File: hdf-cli/cmd/hdf/cmd/query.go
package cmd

import (
	"context"
	"encoding/json"
	"fmt"
	"strings"

	hdfengine "github.com/mitre/hdf-libs/hdf-engine/go/v3"
	"github.com/spf13/cobra"
)

// Global flag variables for query command (used by runQuery).
var (
	queryStatus      []string
	querySeverity    []string
	queryImpact      string
	queryCCI         []string
	queryNIST        []string
	querySTIGID      string
	queryTag         []string
	queryDisposition []string
	queryPoams       string
	querySearch      string
	queryProfile     string
	queryCount       bool
	queryLimit       int
)

// NewQueryCmd creates a new query command with fresh state.
func NewQueryCmd() *cobra.Command {
	// Local flag variables for this command instance
	var (
		localQueryStatus      []string
		localQuerySeverity    []string
		localQueryImpact      string
		localQueryCCI         []string
		localQueryNIST        []string
		localQuerySTIGID      string
		localQueryTag         []string
		localQueryDisposition []string
		localQueryPoams       string
		localQuerySearch      string
		localQueryProfile     string
		localQueryCount       bool
		localQueryLimit       int
	)

	cmd := &cobra.Command{
		Use:   "query <file>",
		Short: "Search and filter requirements in an HDF document",
		Long: `Search and filter requirements based on status, severity, tags, and text.

Different flags are combined with AND logic. Repeating the same flag
uses OR logic within that filter.

--status reports EFFECTIVE status, so a requirement with a governing waiver is
already off "failed" before the filter sees it. --disposition names the type of
the override doing that, and --poams reports whether a remediation plan is still
in force; "none-valid" covers no POA&M, an empty list, and only-lapsed ones
alike, because a plan that has expired is not a plan.

Examples:
  hdf query results.json --status failed
  hdf query results.json --status failed --status not_reviewed
  hdf query results.json --status failed --severity high
  hdf query results.json --severity high --severity critical
  hdf query results.json --cci CCI-000366 --cci CCI-000172
  hdf query results.json --nist "AC-2" --nist "CM-6*"
  hdf query results.json --id V-230221
  hdf query results.json --tag "severity:high" --tag "severity:critical"
  hdf query results.json --search "password"
  hdf query results.json --impact ">0.5" --status failed
  hdf query results.json --baseline "RHEL9-STIG"
  hdf query results.json --disposition waiver --severity critical
  hdf query results.json --status failed --poams none-valid
  hdf query results.json --status failed --count`,
		Args: cobra.ArbitraryArgs,
		RunE: func(cmd *cobra.Command, args []string) error {
			// Sync local flags to global variables for runQuery
			queryStatus = localQueryStatus
			querySeverity = localQuerySeverity
			queryImpact = localQueryImpact
			queryCCI = localQueryCCI
			queryNIST = localQueryNIST
			querySTIGID = localQuerySTIGID
			queryTag = localQueryTag
			queryDisposition = localQueryDisposition
			queryPoams = localQueryPoams
			// An unrecognized value would match nothing and report a clean run,
			// which is the false green this vocabulary exists to avoid. Reject it
			// before any document is read.
			if queryPoams != "" && !hdfengine.ValidPoamFilter(queryPoams) {
				return fmt.Errorf("unknown --poams value %q (expected %q or %q)",
					queryPoams, hdfengine.PoamValid, hdfengine.PoamNoneValid)
			}
			for _, status := range queryStatus {
				if !hdfengine.ValidStatus(status) {
					return fmt.Errorf("unknown --status value %q (expected one of: %s)",
						status, strings.Join(hdfengine.StatusValues, ", "))
				}
			}
			for _, severity := range querySeverity {
				if !hdfengine.ValidSeverity(severity) {
					return fmt.Errorf("unknown --severity value %q (expected one of: %s)",
						severity, strings.Join(hdfengine.SeverityValues, ", "))
				}
			}
			for _, disposition := range queryDisposition {
				if !hdfengine.ValidDisposition(disposition) {
					return fmt.Errorf("unknown --disposition value %q (expected one of: %s)",
						disposition, strings.Join(hdfengine.DispositionValues, ", "))
				}
			}
			querySearch = localQuerySearch
			queryProfile = localQueryProfile
			queryCount = localQueryCount
			queryLimit = localQueryLimit
			files, err := expandGlobs(args)
			if err != nil {
				return err
			}
			if len(files) > 1 {
				return runQueryBulk(cmd, files)
			}
			return runQuery(cmd, args)
		},
	}

	cmd.Flags().StringArrayVarP(&localQueryStatus, "status", "s", nil, "Filter by status (repeatable, OR logic): passed, failed, error, not_applicable, not_reviewed")
	cmd.Flags().StringArrayVar(&localQuerySeverity, "severity", nil, "Filter by severity (repeatable, OR logic): critical, high, medium, low, informational")
	cmd.Flags().StringVar(&localQueryImpact, "impact", "", "Filter by impact (e.g., \">0.5\", \">=0.7\", \"0.5\")")
	cmd.Flags().StringArrayVar(&localQueryCCI, "cci", nil, "Filter by CCI identifier (repeatable, OR logic)")
	cmd.Flags().StringArrayVar(&localQueryNIST, "nist", nil, "Filter by NIST control (repeatable, OR logic; supports globs)")
	cmd.Flags().StringVar(&localQuerySTIGID, "id", "", "Filter by requirement ID, STIG ID, GID, or group title")
	cmd.Flags().StringArrayVarP(&localQueryTag, "tag", "t", nil, "Filter by tag key:value (repeatable, OR logic)")
	cmd.Flags().StringArrayVar(&localQueryDisposition, "disposition", nil,
		"Filter by the governing override's type (repeatable, OR logic): waiver, falsePositive, riskAdjustment, attestation, operationalRequirement, inherited, poam")
	cmd.Flags().StringVar(&localQueryPoams, "poams", "",
		"Filter by remediation-plan validity: valid (a POA&M still in force) or none-valid (none, empty, or only lapsed)")
	cmd.Flags().StringVar(&localQuerySearch, "search", "", "Search in title and description")
	cmd.Flags().StringVarP(&localQueryProfile, "baseline", "p", "", "Filter by profile name")
	cmd.Flags().BoolVarP(&localQueryCount, "count", "c", false, "Show only the count of matching requirements")
	cmd.Flags().IntVarP(&localQueryLimit, "limit", "l", 0, "Limit number of results (0 = unlimited)")

	return cmd
}

func runQuery(_ *cobra.Command, args []string) error {
	var filename string
	if len(args) == 0 || args[0] == "-" {
		filename = "-"
	} else {
		filename = args[0]
	}

	data, err := readInputFile(filename)
	if err != nil {
		return err
	}

	if _, typeErr := requireDocumentType(data, []string{"results"}, "hdf query"); typeErr != nil {
		return typeErr
	}

	results, err := parseHDFResults(data)
	if err != nil {
		return fmt.Errorf("failed to parse HDF file: %w", err)
	}

	if queryImpact != "" && !hdfengine.ValidImpactFilter(queryImpact) {
		return fmt.Errorf("invalid --impact filter %q: use a comparison like >0.5, >=0.7, <0.5, or =0", queryImpact)
	}

	// Filtering is delegated to the shared hdf-engine library; the CLI supplies
	// its display-status resolver so the engine stays convention-agnostic.
	matches := hdfengine.Filter(context.Background(), results, hdfengine.Options{
		Status:      queryStatus,
		Severity:    querySeverity,
		Impact:      queryImpact,
		CCI:         queryCCI,
		NIST:        queryNIST,
		ID:          querySTIGID,
		Tag:         queryTag,
		Disposition: queryDisposition,
		Poams:       queryPoams,
		Search:      querySearch,
		Baseline:    queryProfile,
		Limit:       queryLimit,
		Count:       queryCount,
		StatusOf:    determineControlStatus,
	})

	return outputQueryResults(matches)
}

// outputQueryResults formats and prints query results.
func outputQueryResults(matches []hdfengine.Match) error {
	if queryCount {
		if jsonOutput {
			output, _ := json.Marshal(map[string]int{"count": len(matches)})
			fmt.Println(string(output))
		} else {
			fmt.Println(len(matches))
		}
		return nil
	}

	if jsonOutput {
		output, _ := json.MarshalIndent(matches, "", "  ")
		fmt.Println(string(output))
		if len(matches) == 0 {
			return &exitCodeError{code: 1, message: "no matching requirements"}
		}
		return nil
	}

	// Human-readable output
	if len(matches) == 0 {
		fmt.Println("No matching requirements found.")
		return &exitCodeError{code: 1, message: "no matching requirements"}
	}

	if !noHeaders {
		fmt.Printf("Found %d matching requirement(s):\n\n", len(matches))
	}

	tbl := NewTable(
		Column{Header: "ID"},
		Column{Header: "Status"},
		Column{Header: "Severity"},
		Column{Header: "Title"},
	)
	for _, m := range matches {
		title := sanitizeOutput(m.Title)
		if len(title) > 55 { //nolint:mnd // truncate long titles for table display
			title = title[:52] + "..."
		}
		if title == "" {
			title = "(no title)"
		}
		tbl.AddRow(sanitizeOutput(m.ID), m.Status, severityToLabel(m.Severity), title)
	}
	tbl.Render()

	return nil
}

// Severity constants aligned with CVSS 3.x bands normalized to 0-1.
// Bands: 0.9-1.0=critical, 0.7-0.8=high, 0.4-0.6=medium, 0.1-0.3=low,
// 0.0=informational. The vocabulary is the schema's severity enum, so a filter
// value matches what DeriveSeverity produces.
const (
	SeverityCritical      = "critical"
	SeverityHigh          = "high"
	SeverityMedium        = "medium"
	SeverityLow           = "low"
	SeverityInformational = "informational"
	// SeverityNoneLegacy is the pre-3.7 spelling of informational, still
	// accepted as a filter value so an existing command line keeps working.
	SeverityNoneLegacy = "none"
)

func severityToLabel(severity string) string {
	switch severity {
	case SeverityCritical:
		return "CRIT"
	case SeverityHigh:
		return "HIGH"
	case SeverityMedium:
		return "MED "
	case SeverityLow:
		return "LOW "
	case SeverityInformational:
		return "INFO"
	default:
		return "INFO"
	}
}

func runQueryBulk(cmd *cobra.Command, files []string) error {
	return runBulk(files, "query", "queried", func(file string) error {
		return runQuery(cmd, []string{file})
	})
}

File: hdf-engine/src/query.ts
// Requirements query engine — the TypeScript peer of hdf-engine/go/filter.go.
// A pure filter(results, options) with no shared mutable state, kept at
// behavioural parity with the Go implementation (see test/query.test.ts, which
// runs both over the same fixture).

import type { HDFResults, EvaluatedRequirement } from '@mitre/hdf-schema';
import { governingStatusOverrideIndex, parseTimestamp } from '@mitre/hdf-utilities';
import { deriveSeverity } from './compliance.js';
import { effectiveImpactOf, overrideInputs } from './effective.js';
import { normalizeKey, normalizeFilterValue } from './vocabulary.js';
import { safeGlobMatch } from './safematch.js';

/**
 * FilterOptions configures a requirements query. Every filter input arrives here;
 * distinct filters combine with AND, repeated values within a filter with OR.
 * statusOf resolves a requirement's display status, injected so the engine stays
 * agnostic to the caller's status-string convention (undefined → empty status).
 */
export interface FilterOptions {
  status?: string[];
  severity?: string[];
  /**
   * Compares the EFFECTIVE impact — the governing non-expired impact override's
   * value, else the requirement's own.
   */
  impact?: string;
  /**
   * Compares the requirement's own impact, ignoring overrides. It exists because
   * impact resolves them: the governance policy "an override may not move a
   * critical below 0.7" needs both scores, and no other key reaches the
   * unadjusted one.
   */
  rawImpact?: string;
  cci?: string[];
  nist?: string[];
  id?: string;
  tag?: string[];
  search?: string;
  baseline?: string;
  /**
   * The TYPE of the override that governs the requirement (waiver,
   * falsePositive, riskAdjustment, …), OR across values. Resolved through the
   * same governing-override rule effective status uses rather than read from the
   * stored disposition field, which is an output cache.
   */
  disposition?: string[];
  /**
   * Remediation-plan validity: 'valid' for a requirement carrying a POA&M still
   * in force, 'none-valid' for one carrying none, an empty list, or only lapsed
   * ones. Absence and expiry are one concept on purpose.
   */
  poams?: string;
  /** RFC3339 reference clock for expiry; undefined means now. */
  now?: string;
  limit?: number;
  count?: boolean;
  statusOf?: (control: EvaluatedRequirement) => string;
}

/**
 * A single query result row. baselineIndex and index are the match's position in
 * the result set (results.baselines[baselineIndex].requirements[index]) and are
 * the only unique identity a match has: baseline names repeat in shipped
 * converter output and requirement ids repeat within a baseline, so (baseline,
 * id) is not a key. Address the source requirement by position, never by name.
 */
export interface Match {
  id: string;
  title: string;
  status: string;
  impact: number;
  severity: string;
  baseline: string;
  baselineIndex: number;
  index: number;
}

type FilterFunc = (control: EvaluatedRequirement, status: string, severity: string) => boolean;

/**
 * filter returns the requirements across the result set's baselines that satisfy
 * options. Applies to requirement collections (results/baseline documents); the
 * calling adapter rejects document types that carry no requirements.
 */
export function filter(results: HDFResults, options: FilterOptions): Match[] {
  const filters = buildFilters(options);
  const matches: Match[] = [];

  for (const [baselineIndex, baseline] of (results.baselines ?? []).entries()) {
    if (options.baseline && !matchesGlob(baseline.name, options.baseline)) {
      continue;
    }
    for (const [index, control] of (baseline.requirements ?? []).entries()) {
      if (options.limit && options.limit > 0 && matches.length >= options.limit && !options.count) {
        return matches;
      }

      const status = options.statusOf ? options.statusOf(control) : '';
      // Explicit STIG severity wins; impact-derived only as a fallback — the
      // canonical rule shared with the compliance counts (deriveSeverity), so
      // query rows and compliance never disagree on a requirement's severity.
      // Effective impact for the same reason the impact filter uses it: a
      // governing riskAdjustment moves the requirement into the band it was
      // re-scored into, and filter is an override-aware surface.
      const impact = effectiveImpactOf(control, options.now);
      const severity = deriveSeverity(impact, control.severity ?? null);

      if (!applyFilters(control, status, severity, filters)) {
        continue;
      }

      matches.push({
        id: control.id,
        title: control.title ?? '',
        status,
        // The effective impact, for the same reason status carries the
        // effective status and a Match has no raw twin of either.
        impact,
        severity,
        baseline: baseline.name,
        baselineIndex,
        index,
      });
    }
  }
  return matches;
}

function buildFilters(options: FilterOptions): FilterFunc[] {
  const filters: FilterFunc[] = [];

  // Compared through normalizeKey so the CLI's display spelling and the schema's
  // camelCase are one value: a filter copied from `hdf query` and a rule written
  // against the schema must select the same requirements.
  if (options.status && options.status.length > 0) {
    const statuses = options.status.map((s) => normalizeKey(normalizeFilterValue('status', s)));
    // BOTH sides go through the same alias map: the injected resolver may speak
    // the CLI's display vocabulary while the spec speaks the schema's.
    filters.push((_c, s) => statuses.includes(normalizeKey(normalizeFilterValue('status', s))));
  }

  // Normalized the same way, and through the alias map as well, so the pre-3.7
  // 'none' spelling keeps naming the informational severity on every surface
  // rather than only in the CLI.
  if (options.severity && options.severity.length > 0) {
    const severities = options.severity.map((s) => normalizeKey(normalizeFilterValue('severity', s)));
    filters.push((_c, _s, severity) =>
      severities.includes(normalizeKey(normalizeFilterValue('severity', severity)))
    );
  }

  // Disposition (OR across values). A requirement with no governing override
  // matches nothing, which is what makes "waived" and "not waived" opposites.
  if (options.disposition && options.disposition.length > 0) {
    const wanted = options.disposition.map((d) => normalizeKey(normalizeFilterValue('disposition', d)));
    filters.push((c) => {
      const governing = governingDisposition(c, options.now);
      return governing !== '' && wanted.includes(normalizeKey(normalizeFilterValue('disposition', governing)));
    });
  }

  // POA&M validity. An unrecognized value matches NOTHING rather than silently
  // degrading, matching the impact filter's posture — callers validate with
  // validPoamFilter and reject before filtering.
  if (options.poams) {
    const wantValid = poamFilterWantsValid(options.poams);
    filters.push((c) => wantValid !== undefined && hasValidPoam(c, options.now) === wantValid);
  }

  // Compared against EFFECTIVE impact, so a governing riskAdjustment is
  // honoured. Status has resolved overrides all along; an impact filter that
  // ignored a formal re-score was the same amendments-blindness in the field
  // nobody looked at.
  if (options.impact) {
    const [op, val] = parseImpactFilter(options.impact);
    filters.push((c) => compareImpact(effectiveImpactOf(c, options.now), op, val));
  }

  // The unadjusted twin, same grammar and same safe-degradation.
  if (options.rawImpact) {
    const [op, val] = parseImpactFilter(options.rawImpact);
    filters.push((c) => compareImpact(c.impact, op, val));
  }

  if (options.cci && options.cci.length > 0) {
    const ccis = options.cci.map((c) => c.toUpperCase());
    filters.push((c) => ccis.some((cci) => tagContains(c.tags, 'cci', cci)));
  }

  if (options.nist && options.nist.length > 0) {
    const nist = options.nist;
    filters.push((c) => nist.some((n) => tagMatchesGlob(c.tags, 'nist', n)));
  }

  if (options.id) {
    const id = options.id;
    filters.push(
      (c) =>
        tagContains(c.tags, 'stig_id', id) ||
        tagContains(c.tags, 'gid', id) ||
        tagContains(c.tags, 'gtitle', id) ||
        c.id === id,
    );
  }

  if (options.tag && options.tag.length > 0) {
    const tagFilters: { key: string; value: string }[] = [];
    for (const t of options.tag) {
      const idx = t.indexOf(':');
      if (idx >= 0) {
        tagFilters.push({ key: t.slice(0, idx), value: t.slice(idx + 1) });
      }
    }
    if (tagFilters.length > 0) {
      filters.push((c) => tagFilters.some((tf) => tagMatchesGlob(c.tags, tf.key, tf.value)));
    }
  }

  if (options.search) {
    const search = options.search.toLowerCase();
    filters.push((c) => {
      if (c.id.toLowerCase().includes(search)) return true;
      if (c.title && c.title.toLowerCase().includes(search)) return true;
      for (const desc of c.descriptions ?? []) {
        if (desc.data.toLowerCase().includes(search)) return true;
      }
      return false;
    });
  }

  return filters;
}

function applyFilters(
  control: EvaluatedRequirement,
  status: string,
  severity: string,
  filters: FilterFunc[],
): boolean {
  return filters.every((f) => f(control, status, severity));
}

// DECIMAL_FLOAT is the shared impact-filter operand grammar — a plain decimal:
// optional sign, digits with an optional fraction (or a leading-dot fraction),
// optional decimal exponent. It excludes the forms JS Number() accepts but that
// make no sense as a 0.0–1.0 threshold and diverge from Go strconv.ParseFloat:
// hex/binary/octal integers (0x1f→31, 0b101→5, 0o17→15) and Infinity. The Go
// engine applies the identical pattern so a filter is accepted or rejected the
// same in both languages (bead 4908.15).
const DECIMAL_FLOAT = /^[+-]?(\d+(\.\d*)?|\.\d+)([eE][+-]?\d+)?$/;

// parseDecimal parses a plain-decimal operand, returning null for anything
// outside the shared grammar (including overflow to Infinity, e.g. '1e400', so
// Go's ParseFloat ErrRange rejection is mirrored).
function parseDecimal(s: string): number | null {
  if (!DECIMAL_FLOAT.test(s)) return null;
  const v = Number(s);
  return Number.isFinite(v) ? v : null;
}

// An invalid operand safe-degrades to ('=', NaN), which compareImpact treats as
// matching NOTHING — mirroring the Go engine's `return ok && compareImpact(...)`.
// It must never coerce to ('=', 0), which silently returns confidently-wrong
// impact==0 rows for a typo'd or garbage filter.
export function parseImpactFilter(f: string): [string, number] {
  const trimmed = f.trim();
  for (const op of ['>=', '<=', '>', '<', '=']) {
    if (trimmed.startsWith(op)) {
      const v = parseDecimal(trimmed.slice(op.length).trim());
      return v === null ? ['=', NaN] : [op, v];
    }
  }
  const v = parseDecimal(trimmed);
  return v === null ? ['=', NaN] : ['=', v];
}

export function compareImpact(impact: number, op: string, val: number): boolean {
  switch (op) {
    case '>':
      return impact > val;
    case '>=':
      return impact >= val;
    case '<':
      return impact < val;
    case '<=':
      return impact <= val;
    case '=':
      return impact === val;
    default:
      return false;
  }
}

export function tagContains(tags: Record<string, unknown>, key: string, value: string): boolean {
  const tagVal = tags?.[key];
  if (tagVal === undefined) {
    return false;
  }
  const want = value.toLowerCase();
  if (typeof tagVal === 'string') {
    return tagVal.toLowerCase() === want;
  }
  if (Array.isArray(tagVal)) {
    return tagVal.some((item) => typeof item === 'string' && item.toLowerCase() === want);
  }
  return false;
}

export function tagMatchesGlob(tags: Record<string, unknown>, key: string, pattern: string): boolean {
  const tagVal = tags?.[key];
  if (tagVal === undefined) {
    return false;
  }
  if (typeof tagVal === 'string') {
    return safeGlobMatch(tagVal, pattern);
  }
  if (Array.isArray(tagVal)) {
    return tagVal.some((item) => typeof item === 'string' && safeGlobMatch(item, pattern));
  }
  return false;
}

/** matchesGlob reports whether s matches the glob pattern (case-insensitive). */
export function matchesGlob(s: string, pattern: string): boolean {
  return safeGlobMatch(s, pattern);
}

/**
 * The closed vocabulary the disposition filter accepts: the schema's
 * Override_Type enum. A value outside it can only ever match nothing, which
 * would report a clean run over a filter the caller believed was applied.
 * Parity: DispositionValues in go/filter.go, pinned by the shared case table.
 * Membership is not pinned to the schema: an eighth override type would be
 * silently unfilterable in both languages until someone adds it here.
 */
export const DISPOSITION_VALUES = [
  'waiver',
  'attestation',
  'poam',
  'inherited',
  'falsePositive',
  'riskAdjustment',
  'operationalRequirement',
] as const;

/**
 * Reports whether s names an override type. Callers validate with this and
 * reject, rather than letting a typo match nothing and pass — the same contract
 * validPoamFilter provides for the other amendment filter.
 */
// Re-exported from vocabulary.ts, where all four validators share one canonical()
// rather than this one inferring "is an alias" from "the normalizer changed it".
export { validDisposition } from './vocabulary.js';

/**
 * The two values the poams filter accepts. 'none-valid' covers a requirement
 * with no POA&M, with an empty list, and with only lapsed ones: a plan that has
 * expired is not a plan, and splitting the two cases would create a value whose
 * only use is to be chosen by mistake.
 */
export const POAM_VALID = 'valid';
export const POAM_NONE_VALID = 'none-valid';

/**
 * Reports whether s is a POA&M validity value the filter understands. Callers
 * validate with this and reject, rather than letting an unrecognized value match
 * nothing and pass.
 */
export function validPoamFilter(s: string): boolean {
  return poamFilterWantsValid(s) !== undefined;
}

function poamFilterWantsValid(s: string): boolean | undefined {
  switch (s.trim().toLowerCase()) {
    case POAM_VALID:
      return true;
    case POAM_NONE_VALID:
      return false;
    default:
      return undefined;
  }
}

/**
 * The type of the override that governs the requirement, or '' when none does.
 * The index comes from the shared governing-override rule — most recently
 * applied, non-expired, carrying a status — so disposition and effective status
 * can never disagree about which override is in force.
 */
function governingDisposition(control: EvaluatedRequirement, now?: string): string {
  const overrides = control.statusOverrides ?? [];
  const index = governingStatusOverrideIndex(overrideInputs(control), now);
  return index < 0 ? '' : (overrides[index]?.type ?? '');
}

/**
 * Whether the requirement carries a POA&M still in force. expiresAt is required
 * by the schema, so a POA&M always has a deadline to judge; one exactly at the
 * reference instant has passed, matching how an override's expiry is judged.
 */
// Note the inverted zero: hdfutil treats an override with no expiresAt as never
// expiring, while a POA&M without a deadline is treated as already lapsed. The
// schema requires poams[].expiresAt so it should not arise, but the two
// same-shaped fields mean opposite things when empty.
function hasValidPoam(control: EvaluatedRequirement, now?: string): boolean {
  // parseTimestamp returns null on an unparseable value; falling back to the
  // wall clock there is wrong in the other direction, so an unusable reference
  // is treated as no reference and the caller's own validation catches it.
  const parsed = now ? parseTimestamp(now) : null;
  const ref = parsed ? parsed.getTime() : Date.now();
  return (control.poams ?? []).some((poam) => new Date(poam.expiresAt).getTime() > ref);
}
